### PR TITLE
ath79: add support for D-Link DIR-869 A1

### PR DESCRIPTION
**Pull request by AvdWerfhorst**
Signed-off-by: AvdWerfhorst <a.werfhorst@alwe.nl>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

### CODE REVIEW

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -144,6 +144,10 @@ dlink,dap-1365-a1)
 dlink,dir-859-a1)
 	ucidef_set_led_switch "internet" "WAN" "$boardname:green:internet" "switch0" "0x20"
 	;;
+dlink,dir-869-a1)
+	ucidef_set_led_switch "power" "WAN" "$boardname:orange:power" "switch0" "0x20"
```
Comment (adrianschmutzler) [inline target/linux/ath79/generic/base-files/etc/board.d/01_leds:148]: really?

**Comment by adrianschmutzler** [comment]
Please add specifications, flashing instructions and an overview of MAC addresses used in the vendor firmware to the commit message, similar to previous submissions for similar devices.

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		reset {
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
```
Comment (adrianschmutzler) [inline target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:49]: Are those really swapped compared to the dir-859?

**Comment by adrianschmutzler** [comment]
Please create a shared DTSI with the dir-859-a1, similar to how it's done for the dir-842-cs ...

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts (hunk 2 above)
Comment (AvdWerfhorst) [reply target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:49]: I hope not to sound too much like a noob. This is my first try into coding with github at all. So there are probably a dozen beginners mistakes.

> Please add specifications, flashing instructions and an overview of MAC addresses used in the vendor firmware to the commit message, similar to previous submissions for similar devices.

Can you please point me in the right direction for this? I guess you also mean the hardware specs, instead of just MAC addresses. I did see some commits with digits added for certain interfaces. I guess that is what you mean, what digit is used for which interface. Am I right about this?

I did not use the vendor firmware for creating this patch. I started by interpreting the ar71xx source, putting that next to the dir-859 source and basically used the dir-859 as a template, where I added dir-869 values where needed. 

> Are those really swapped compared to the dir-859?

Yes they are. I took my gpio values from this file. https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ar71xx/files/arch/mips/ath79/mach-dir-869-a1.c;h=b0894e37cce9135d37ebf12847d8685d048f4ae1;hb=c4c72a9a42e07bed8f9e9eb214c0debfb5d7ab90



> Please create a shared DTSI with the dir-859-a1, similar to how it's done for the dir-842-cs ...

I have to look into that. But I guess that will be manageable.

**Comment by aparcar** [comment]
Your patch has multiple formal issues, please check [here](https://gitlab.com/aparcar/openwrt/-/pipelines/177256243)

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts (hunk 2 above)
Comment (adschm) [reply target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:49]: Commit message:

Just look at what people did before you (preferably in recent patches), e.g.
https://github.com/openwrt/openwrt/commit/68ac3f2cddab8422d7de0ce1a78d23edf29012e7

> I did not use the vendor firmware for creating this patch. I started by interpreting the ar71xx source, putting that next to the dir-859 source and basically used the dir-859 as a template, where I added dir-869 values where needed.

Then we will have to live with that. In the past, many people just invented their own MAC address assignment scheme, so actually checking with OEM firmware might be helpful to find the true assignment. However, that's optional if there is ar71xx setup already.

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+		led_power: power {
+			label = "dir-869-a1:orange:power";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		internet {
+			label = "dir-869-a1:white:internet";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
```
Comment (adschm) [inline target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:32]: Just looked at the manuel:
https://media.dlink.eu/ftp/products/dir/dir-869/documentation/DIR-869_man_revA_V1-02_eu_multi_20170516.pdf

So, this actually appears to be a power LED with two colors, so please use it like that. Based on the manual (as far as possible):

```
	aliases {
		led-boot = &led_power_orange;
		led-failsafe = &led_power_orange;
		led-running = &led_power_white;
		led-upgrade = &led_power_white;
	};

	leds {
		compatible = "gpio-leds";

		led_power_orange: power_orange {
			label = "dir-869-a1:orange:power";
			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
		};

		led_power_white: power_white {
			label = "dir-869-a1:white:power";
			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
		};
```

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts (hunk 2 above)
Comment (adschm) [reply target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:49]: Mode switch is missing here.

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		phy-mode = "sgmii";
+		qca,mib-poll-interval = <500>;
+
+		qca,ar8327-initvals = <
+			0x04 0x00080080 /* PORT0 PAD MODE CTRL */
+			0x10 0x81000080 /* POWER_ON_STRAP */
+			0x50 0xcc35cc35 /* LED_CTRL0 */
+			0x54 0xcb37cb37 /* LED_CTRL1 */
+			0x58 0x00000000 /* LED_CTRL2 */
+			0x5c 0x00f3cf00 /* LED_CTRL3 */
```
Comment (adschm) [inline target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:142]: Do you have LAN LEDs?

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds [hunk 5]
```diff
@@ -144,6 +144,10 @@ dlink,dap-1365-a1)
...
+	ucidef_set_led_switch "internet" "WAN" "$boardname:white:internet" "switch0" "0x20"
+	;;
```
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/01_leds:150]: Based on my finding above, all changes in this file can be dropped.

File: target/linux/ath79/generic/base-files/etc/board.d/02_network [hunk 6]
```diff
@@ -428,6 +429,7 @@ ath79_setup_macs()
 		label_mac=$lan_mac
 		;;
 	dlink,dir-859-a1|\
+	dlink,dir-869-a1|\
```
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/02_network:432]: Does the device have a label with a MAC address on it?

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	gpio-export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		gpio_switch_reset {
+			gpio-export,name = "dir-869-a1:reset:switch";
+			gpio-export,output = <1>;
+			gpios = <&gpio 11 GPIO_ACTIVE_HIGH>;
+		};
+	};
```
Comment (s-2) [inline target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:61]: We don't need gpio-export anymore to declare the reset pin for the external switch, the driver has the option to specify this in the `phy0` node in the dts:
  `reset-gpios = <&gpio 11 GPIO_ACTIVE_LOW>;`
It is active low here, since the driver will write the reset output to logic `0` (i.e. "not reset"), which will result in a high level on the line (as it is currently achieved by writing `1` to an active high gpio).

Please see my PR #3027 for reference, this could be changed for DIR-859 as well, I just didn't have a device for testing at that time...

File: target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <30000000>;
```
Comment (s-2) [inline target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:80]: This can be increased to `50000000`.

**Comment by s-2** [comment]
I think we can even base both devices on `qca9563_dlink_dir-842-c.dtsi`, since most of the stuff that was fixed in PR #3027 (except status LED GPIO) would apply to DIR-859 as well.

Considering that DIR-842 is the only device still active (while DIR-859 and DIR-869 are end-of-life), I would not even bother to change the `.dtsi` filename to something generic, besides there are many models names in the DIR-8xx range that use other chipsets and/or firmware formats, so any generic name would rather add confusion.

Basically, DIR-859 is just the AC1750 variant of DIR-842, while DIR-869 is the 1000mW AC1750 variant (that comes in a weird case)...

File: target/linux/ath79/image/generic.mk [hunk 9]
```diff
@@ -752,6 +752,18 @@ define Device/dlink_dir-859-a1
...
 
+define Device/dlink_dir-869-a1
+  $(Device/seama)
+  SOC := qca9563
+  DEVICE_VENDOR := D-Link
+  DEVICE_MODEL := DIR-869
+  DEVICE_VARIANT := A1
+  IMAGE_SIZE := 15872k
+  DEVICE_PACKAGES :=  kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
+  SEAMA_SIGNATURE := wrgac54_dlink.2015_dir869
```
Comment (s-2) [inline target/linux/ath79/image/generic.mk:763]: Please add the line
`  SUPPORTED_DEVICES += dir-869-a1`
here, since this device has been supported in ar71xx, but the format of device names was changed for ath79.

This will add the old model string from ar71xx to the metadata appended to `.sysupgrade` images, allowing routers currently running ar71xx to update smoothly to ath79.

Just updated mine from LEDE, great to see this device ported to ath79! :slightly_smiling_face:

**Comment by adschm** [comment]
This is abandoned by the author (looks like he never updated it at all), so I'm closing it.
